readme.txt
=== WooCommerce Composite Products - Conditional Components ===

Contributors: franticpsyx, SomewhereWarm
Tags: woocommerce, composite, products, conditional, components, scenario, action, custom
Requires at least: 3.8
Tested up to: 4.4
Stable tag: 1.1.1
WC requires at least: 2.3
WC tested up to: 2.5
License: GNU General Public License v3.0
License URI: http://www.gnu.org/licenses/gpl-3.0.html

Adds a custom "Hide Components" Scenario action for conditionally hiding Composite Components. Requires WooCommerce Composite Products.

== Description ==

[Composite Products](http://www.woothemes.com/products/composite-products/) is an official [WooCommerce](http://www.woothemes.com/woocommerce/) extension suitable for creating dynamic product kits made up of multiple components, such as custom computers, skateboards, gift-boxes and made-to-order jewellery. Composite Products includes **Scenarios**, a powerful, developer-oriented feature that allows you to associate custom **actions** with specific configuration **states** of a Composite Product. By default, Scenarios can be used to conditionally **enable or disable specific Component Options** at every step of the configuration, depending on previous product/variation selections. You can find out more about Scenarios by taking a look at [the documentation](http://docs.woothemes.com/document/composite-products/) of the extension.

**Composite Products - Conditional Components** is a Composite Products plugin that relies on Scenarios and allows you to **conditionally hide one or more Components** depending on product/variation selections. The plugin works by adding a new **Hide Components** Scenario action.

Developers may checkout and contribute to the plugin at its [GitHub Repository](http://github.com/somewherewarm/woocommerce-composite-products-conditional-components).

**Important**: Before using this plugin ensure that you have installed the latest versions of both **WooCommerce** and **WooCommerce Composite Products**.

== Installation ==

1. Install the latest versions of WooCommerce and WooCommerce Composite Products.
2. Upload the plugin to the **/wp-content/plugins/** directory.
3. Activate the plugin through the **Plugins** menu in WordPress.


changelog.txt
*** WooCommerce Composite Products - Conditional Components ***

2016.04.04 - version 1.1.1
 * Fix - Componentized layout support.

2016.03.05 - version 1.1.0
 * Important - Added Composite Products v3.6.0 support. Composite Products v3.6.0 required.

2016.01.09 - version 1.0.3
 * Important - Added Composite Products v3.5.1 support. Composite Products v3.5.1 required.
 * Fix - Take all Composite steps/components into account when calculating active "Hide Components" scenarios.

2015.09.24 - version 1.0.2
 * Tweak - Move tooltips after input fields.

2015.05.29 - version 1.0.1
 * Fix - Add missing scenario filters.

2015.05.20 - version 1.0.0
 * Initial Release.


assets/js/wccp-conditional-components-add-to-cart.js
/* global wc_cp_composite_scripts */

;( function ( $, window, document, undefined ) {

	var conditional_component_scripts = {};

	$( '.composite_data' )

		.on( 'wc-composite-initializing', function() {

			var container_id = $( this ).data( 'container_id' );

			if ( typeof( conditional_component_scripts[ container_id ] ) !== 'undefined' ) {
				return true;
			} else if ( typeof( wc_cp_composite_scripts[ container_id ] ) === 'undefined' ) {
				return false;
			} else {

				var composite = wc_cp_composite_scripts[ container_id ];

				conditional_component_scripts[ container_id ] = new WC_CP_Conditional_Components( composite );

				conditional_component_scripts[ container_id ].init();
			}

		} );

	function WC_CP_Conditional_Components ( composite ) {

		var script = this;

		this.init = function() {

			/*
			 * Init conditional visibility models.
			 */
			composite.actions.add_action( 'initialize_composite', function() {
				script.init_models();
			}, 21, this );

			/*
			 * Setup filters.
			 */
			composite.actions.add_action( 'initialize_composite', function() {
				script.add_filters();
			}, 21, this );

			/*
			 * Init conditional visibility views.
			 */
			composite.actions.add_action( 'initialize_composite', function() {
				script.init_views();
			}, 51, this );
		};

		/**
		 * Init visibility models.
		 */
		this.init_models = function() {

			var Step_Visibility_Model = function( component, opts ) {

				var self = component;

				/**
				 * Controls the visibility state of a component.
				 */
				var Model = Backbone.Model.extend( {

					initialize: function() {

						var params = {
							is_hidden: false,
						};

						this.set( params );

						if ( self.is_component() ) {

							/**
							 * Update model state when the active scenarios change.
							 */
							composite.actions.add_action( 'active_scenarios_changed', this.update_visibility_state, 15, this );

							/**
							 * Run 'step_visibility_changed' action.
							 */
							this.on( 'change:is_hidden', this.do_action, this );
						}
					},

					do_action: function() {

						/**
						 * Action 'step_visibility_changed':
						 *
						 * @hooked Step_Visibility_View::step_visibility_changed_handler - 10
						 */
						composite.actions.do_action( 'step_visibility_changed', [ self ] );
					},

					update_visibility_state: function() {

						var scenarios        = composite.scenarios.get_active_scenarios(),
							active_scenarios = composite.scenarios.filter_scenarios_by_type( scenarios, 'conditional_components' ),
							hide_component   = false;

						composite.console_log( 'debug:models', '\nUpdating "' + self.get_title() + '" visibility...' );
						composite.console_log( 'debug:models', 'Active "Hide Components" Scenarios: ' + active_scenarios.toString() );

						// Get conditional components data.
						var conditional_components = composite.scenarios.get_scenario_data().scenario_settings.conditional_components;

						// Find if the component is hidden in the active scenarios.
						if ( active_scenarios.length > 0 && typeof( conditional_components ) !== 'undefined' ) {

							// Set hide status.
							$.each( conditional_components, function( scenario_id, hidden_components ) {

								if ( _.contains( active_scenarios, scenario_id.toString() ) ) {
									if ( _.contains( hidden_components, self.component_id.toString() ) ) {
										hide_component = true;
									}
								}
							} );
						}

						composite.console_log( 'debug:models', '\nUpdating \'Step_Visibility_Model\': "' + self.get_title() + '", Attribute: "is_hidden": ' + hide_component.toString() );

						composite.debug_tab_count = composite.debug_tab_count + 2;
						this.set( { is_hidden: hide_component } );
						composite.debug_tab_count = composite.debug_tab_count - 2;
					}

				} );

				var obj = new Model( opts );
				return obj;
			};

			/*
			 * Init visibility models.
			 */
			$.each( composite.get_steps(), function( index, component ) {
				component.step_visibility_model = new Step_Visibility_Model( component );
			} );
		};

		/*
		 * Add filters:
		 *
		 * - Step access model state filter.
		 * - Step validation result filter.
		 * - Component 'is_optional' method filter.
		 * - Component validation messages filter.
		 */
		this.add_filters = function() {

			composite.filters.add_filter( 'step_is_valid', this.step_is_valid, 10, this );
			composite.filters.add_filter( 'step_is_locked', this.step_is_locked, 10, this );
			composite.filters.add_filter( 'component_is_optional', this.component_is_optional, 10, this );
			composite.filters.add_filter( 'step_validation_messages', this.step_validation_messages, 10, this );
		};

		/*
		 * Step validation messages filter.
		 */
		this.step_validation_messages = function( messages, context, step ) {

			if ( step.is_component() && step.step_visibility_model.get( 'is_hidden' ) ) {
				messages = [];
			}

			return messages;
		};

		/*
		 * Step validation result filter.
		 */
		this.step_is_valid = function( is_valid, step ) {

			if ( step.is_component() && step.step_visibility_model.get( 'is_hidden' ) ) {
				is_valid = true;
			}

			return is_valid;
		};

		/*
		 * Step access model state filter.
		 */
		this.step_is_locked = function( is_locked, step ) {

			if ( step.is_component() && step.step_visibility_model.get( 'is_hidden' ) ) {
				is_locked = true;
			}

			return is_locked;
		};

		/**
		 * Component 'is_optional' filter: Hidden components are optional.
		 */
		this.component_is_optional = function( is_optional, component ) {

			if ( component.step_visibility_model.get( 'is_hidden' ) ) {
				is_optional = true;
			}

			return is_optional;
		};

		/**
		 * Init visibility views.
		 */
		this.init_views = function() {

			var Step_Visibility_View = function( component, opts ) {

				var self = component;

				var View = Backbone.View.extend( {

					initialize: function() {

						/**
						 * Change prev/next pointers and visibility to skip hidden components:
						 *
						 * - When the visibility model state of a prev/next component changes.
						 * - When the active step changes and prev/next pointers need to be refreshed.
						 */
						composite.actions.add_action( 'step_visibility_changed', this.step_visibility_changed_handler, 10, this );
						composite.actions.add_action( 'active_step_changed', this.update_pointers, 100, this );
					},

					step_visibility_changed_handler: function( step ) {

						/**
						 * Change prev/next component pointers.
						 */
						this.update_pointers();

						/**
						 * Hide prev/next components depending on their visibility model state.
						 */
						if ( step.step_index === self.step_index ) {
							this.update_visibility();
							this.update_views();
						}

						/**
						 * Refresh numeric indexes in component markup.
						 */
						this.update_indexes();
					},

					update_indexes: function() {

						composite.console_log( 'debug:views', '\nRefreshing "' + self.get_title() + '" indexes...' );

						var hidden_count          = 0,
							summary_columns       = composite.$composite_summary.data( 'columns' ),
							summary_element_class = '',
							count                 = 1;

						// Count number of hidden components before this one.
						hidden_count = _.filter( composite.get_steps(), function( component ) {
							if ( component.step_visibility_model.get( 'is_hidden' ) && component.step_index < self.step_index ) {
								return component;
							}
						} ).length;

						count = self.step_index - hidden_count + 1;

						if ( ( ( count - 1 ) % summary_columns ) === 0 || summary_columns === 1 ) {
							summary_element_class = 'first';
						}

						if ( count % summary_columns === 0 ) {
							summary_element_class += ' last';
						}

						// Refresh index in step title.
						self.$step_title.find( '.step_index' ).html( count );

						// Refresh index in composite pagination element.
						if ( false !== composite.composite_pagination_view ) {
							composite.composite_pagination_view.view_elements[ self.step_id ].$pagination_element.find( '.element_index' ).html( count );
						}

						// Refresh index in composite summary elements.
						if ( self.is_component() ) {
							if ( false !== composite.composite_summary_view ) {
								composite.composite_summary_view.view_elements[ self.step_id ].$summary_element.removeClass( 'first last' ).addClass( summary_element_class );
								composite.composite_summary_view.view_elements[ self.step_id ].$summary_element.find( '.step_index' ).html( count );
							}

							$.each( composite.composite_summary_widget_views, function( index, widget_view ) {
								widget_view.composite_summary_view.view_elements[ self.step_id ].$summary_element.find( '.step_index' ).html( count );
							} );
						}
					},

					update_visibility: function() {

						var view = this;

						composite.console_log( 'debug:views', '\nRendering "' + self.get_title() + '" visibility...' );

						if ( false !== composite.composite_pagination_view ) {
							if ( this.model.get( 'is_hidden' ) ) {
								composite.composite_pagination_view.view_elements[ self.step_id ].$pagination_element.hide();
							} else {
								composite.composite_pagination_view.view_elements[ self.step_id ].$pagination_element.show();
							}
						}

						if ( false !== composite.composite_summary_view ) {
							if ( this.model.get( 'is_hidden' ) ) {
								composite.composite_summary_view.view_elements[ self.step_id ].$summary_element.hide();
							} else {
								composite.composite_summary_view.view_elements[ self.step_id ].$summary_element.show();
							}
						}

						$.each( composite.composite_summary_widget_views, function( index, widget_view ) {
							if ( view.model.get( 'is_hidden' ) ) {
								if ( composite.is_initialized ) {
									widget_view.composite_summary_view.view_elements[ self.step_id ].$summary_element.slideUp( 250 );
								} else {
									widget_view.composite_summary_view.view_elements[ self.step_id ].$summary_element.hide();
								}
							} else {
								if ( composite.is_initialized ) {
									widget_view.composite_summary_view.view_elements[ self.step_id ].$summary_element.slideDown( 250 );
								} else {
									widget_view.composite_summary_view.view_elements[ self.step_id ].$summary_element.show();
								}
							}
						} );

						if ( composite.settings.layout !== 'paged' ) {
							if ( this.model.get( 'is_hidden' ) ) {
								self.$el.hide();
							} else {
								self.$el.show();
							}

							if ( self.is_component() ) {
								self.component_title_view.render_navigation_state();
							}
						}
					},

					update_views: function() {

						if ( false !== composite.composite_pagination_view ) {
							composite.composite_pagination_view.render_element_state( self );
						}

						if ( false !== composite.composite_summary_view ) {
							composite.composite_summary_view.render_element_state( self );
						}

						$.each( composite.composite_summary_widget_views, function( index, widget_view ) {
							widget_view.composite_summary_view.render_element_state( self );
						} );
					},

					update_pointers: function() {

						if ( composite.settings.layout_variation === 'componentized' ) {
							return false;
						}

						var curr_step     = composite.get_current_step(),
							next_step_pre = composite.get_next_step(),
							prev_step_pre = composite.get_previous_step();

						if ( self.step_index === curr_step.step_index + 1 ) {

							// If hidden, find next closest that isn't.
							if ( self.step_visibility_model.get( 'is_hidden' ) ) {

								$.each( composite.get_steps(), function( step_index, step ) {
									if ( step_index > self.step_index ) {
										if ( step.is_review() || false === step.step_visibility_model.get( 'is_hidden' ) ) {

											var next_step_new = step;

											composite.console_log( 'debug:views', '\nChanged next component pointers to "' + step.get_title() + '".' );

											if ( false !== next_step_pre ) {
												next_step_pre._is_next = false;
												next_step_pre.$el.removeClass( 'next' );
											}

											next_step_new._is_next = true;
											next_step_new.$el.addClass( 'next' );

											return false;
										}
									}
								} );

							// If not hidden, attempt to reset.
							} else {
								if ( self.step_index !== next_step_pre.step_index ) {

									composite.console_log( 'debug:views', '\nReset next component pointers to "' + self.get_title() + '".' );

									if ( false !== next_step_pre ) {
										next_step_pre.$el.removeClass( 'next' );
										next_step_pre._is_next = false;
									}

									self.$el.addClass( 'next' );
									self._is_next = true;
								}
							}
						}

						if ( self.step_index === curr_step.step_index - 1 ) {

							// If hidden, find next closest that isn't.
							if ( self.step_visibility_model.get( 'is_hidden' ) ) {

								$.each( _.clone( composite.get_steps() ).reverse(), function( index, step ) {
									if ( step.step_index < self.step_index ) {
										if ( step.is_review() || false === step.step_visibility_model.get( 'is_hidden' ) ) {

											var prev_step_new = step;

											composite.console_log( 'debug:views', '\nChanged previous component pointers to "' + step.get_title() + '".' );

											if ( false !== prev_step_pre ) {
												prev_step_pre._is_previous = false;
												prev_step_pre.$el.removeClass( 'prev' );
											}

											prev_step_new._is_previous = true;
											prev_step_new.$el.addClass( 'prev' );

											return false;
										}
									}
								} );

							// If not hidden, attempt to reset.
							} else {
								if ( self.step_index !== prev_step_pre.step_index ) {

									composite.console_log( 'debug:views', '\nReset previous component pointers to "' + self.get_title() + '".' );

									if ( false !== prev_step_pre ) {
										prev_step_pre.$el.removeClass( 'prev' );
										prev_step_pre._is_previous = false;
									}

									self.$el.addClass( 'prev' );
									self._is_previous = true;
								}
							}
						}
					}

				} );

				var obj = new View( opts );
				return obj;
			};

			/*
			 * Init visibility views.
			 */
			$.each( composite.get_steps(), function( index, component ) {
				component.step_visibility_view = new Step_Visibility_View( component, { el: component.$el, model: component.step_visibility_model } );
			} );

			/*
			 * Prevent hidden component selections from being posted.
			 */
			composite.composite_add_to_cart_button_view.$el_button.on( 'click', function() {

				$.each( composite.get_components(), function( index, component ) {

					if ( component.step_visibility_model.get( 'is_hidden' ) ) {
						component.$component_options_select.val( '' );
					}
				} );
			} );

		};
	}

} ) ( jQuery, window, document );


woocommerce-composite-products-conditional-components.php
<?php
/*
* Plugin Name: WooCommerce Composite Products - Conditional Components
* Plugin URI: http://www.woothemes.com/products/composite-products/
* Description: Adds a custom "Hide Components" Scenario action that can be used to conditionally hide Composite Components. Requires WooCommerce Composite Products.
* Version: 1.1.1
* Author: SomewhereWarm
* Author URI: http://somewherewarm.net/
* Developer: Manos Psychogyiopoulos
* Developer URI: http://somewherewarm.net/
*
* Text Domain: woocommerce-composite-products-conditional-components
* Domain Path: /languages/
*
* Requires at least: 3.8
* Tested up to: 4.4
*
* Copyright: © 2009-2015 Manos Psychogyiopoulos.
* License: GNU General Public License v3.0
* License URI: http://www.gnu.org/licenses/gpl-3.0.html
*/

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

class WC_CP_Scenario_Action_Conditional_Components {

	public static $version        = '1.1.1';
	public static $req_cp_version = '3.6.0';

	public static function plugin_url() {
		return plugins_url( basename( plugin_dir_path(__FILE__) ), basename( __FILE__ ) );
	}

	public static function plugin_path() {
		return untrailingslashit( plugin_dir_path( __FILE__ ) );
	}

	public static function init() {
		// Load plugin.
		add_action( 'plugins_loaded', __CLASS__ . '::load' );
	}

	public static function load() {

		global $woocommerce_composite_products;

		if ( ! empty( $woocommerce_composite_products ) && version_compare( $woocommerce_composite_products->version, self::$req_cp_version ) < 0 ) {
			add_action( 'admin_notices', __CLASS__ . '::cp_check_admin_notice' );
			return false;
		}

		// Admin script.
		add_filter( 'woocommerce_composite_script_dependencies', __CLASS__ . '::add_to_cart_script' );

		// Front-end script (where the magic happens).
		add_action( 'admin_enqueue_scripts', __CLASS__ . '::admin_scripts' );

		// Add qty data in scenarios.
		add_filter( 'woocommerce_composite_scenario_data', __CLASS__ . '::scenario_data', 10, 4 );

		// Add 'Hide Components' action in Scenarios.
		add_action( 'woocommerce_composite_scenario_admin_actions_html', __CLASS__ . '::actions_config', 15, 4 );

		// Save 'Hide Components' action settings.
		add_filter( 'woocommerce_composite_process_scenario_data', __CLASS__ . '::actions_save', 10, 5 );

		// Validation.
		add_filter( 'woocommerce_composite_validation_component_is_mandatory', __CLASS__ . '::validate', 10, 6 );
	}

	/**
	 * Displays a warning message if CP version check fails.
	 *
	 * @return string
	 */
	public static function cp_check_admin_notice() {

		echo '<div class="error"><p>' . sprintf( __( '<strong>WooCommerce Composite Products - Conditional Components</strong> requires Composite Products version %s or higher. Please update <strong>WooCommerce Composite Products</strong>.', 'woocommerce-composite-products-conditional-components' ), self::$req_cp_version ) . '</p></div>';
	}

	/**
	 * Οverride notice for empty selection when component is hidden.
	 *
	 * @param  boolean $is_mandatory
	 * @param  string  $component_id
	 * @param  array   $validation_data
	 * @param  array   $common_scenarios
	 * @param  array   $scenario_data
	 * @param  string  $product_id
	 * @return boolean
	 */
	public static function validate( $is_mandatory, $component_id, $validation_data, $common_scenarios, $scenario_data, $product_id ) {

		$conditional_components_scenarios = WC_CP_Scenarios::filter_scenarios_by_type( $common_scenarios, 'conditional_components', $scenario_data );

		if ( $conditional_components_scenarios ) {

			if ( ! empty( $scenario_data[ 'scenario_settings' ][ 'conditional_components' ] ) ) {

				$conditional_components = $scenario_data[ 'scenario_settings' ][ 'conditional_components' ];

				foreach ( $conditional_components as $scenario_id => $hidden_components ) {

					if ( in_array( $scenario_id, $conditional_components_scenarios ) ) {

						if ( ! empty( $hidden_components ) && in_array( $component_id, $hidden_components ) ) {
							$is_mandatory = false;
							break;
						}
					}
				}
			}
		}

		return $is_mandatory;
	}

	/**
	 * Add conditional components data in scenarios.
	 *
	 * @param  array  $scenario_data
	 * @param  array  $scenario_meta
	 * @param  array  $composite_data
	 * @param  mixed  $composite
	 * @return array
	 */
	public static function scenario_data( $scenario_data, $scenario_meta, $composite_data, $composite ) {

		if ( ! empty( $scenario_meta ) ) {

			$hidden_components_settings = array();

			foreach ( $scenario_meta as $scenario_id => $scenario_metadata ) {

				if ( isset( $scenario_metadata[ 'scenario_actions' ][ 'conditional_components' ][ 'is_active' ] ) && $scenario_metadata[ 'scenario_actions' ][ 'conditional_components' ][ 'is_active' ] === 'yes' ) {

					if ( ! empty( $scenario_metadata[ 'scenario_actions' ][ 'conditional_components' ][ 'hidden_components' ] ) ) {

						$hidden_components_settings[ $scenario_id ] = $scenario_metadata[ 'scenario_actions' ][ 'conditional_components' ][ 'hidden_components' ];
					}
				}
			}

			if ( ! empty( $hidden_components_settings ) ) {
				$scenario_data[ 'scenario_settings' ][ 'conditional_components' ] = $hidden_components_settings;
			}

		} else {
			$scenario_data[ 'scenario_settings' ][ 'conditional_components' ] = array();
		}

		return $scenario_data;
	}

	/**
	 * Save conditional components settings in scenarios.
	 *
	 * @param  array  $scenario_meta
	 * @param  array  $scenario_post_data
	 * @param  string $scenario_id
	 * @param  array  $composite_meta
	 * @param  string $composite_id
	 * @return array
	 */
	public static function actions_save( $scenario_meta, $scenario_post_data, $scenario_id, $composite_meta, $composite_id ) {

		global $woocommerce_composite_products;

		$composite = wc_get_product( $composite_id );

		// Save active state
		if ( ! empty( $scenario_post_data[ 'scenario_actions' ][ 'conditional_components' ][ 'is_active' ] ) ) {
			$scenario_meta[ 'scenario_actions' ][ 'conditional_components' ][ 'is_active' ] = 'yes';
		} else {
			$scenario_meta[ 'scenario_actions' ][ 'conditional_components' ][ 'is_active' ] = 'no';
		}

		// Save settings
		if ( ! empty( $scenario_post_data[ 'scenario_actions' ][ 'conditional_components' ][ 'hidden_components' ] ) ) {
			$scenario_meta[ 'scenario_actions' ][ 'conditional_components' ][ 'hidden_components' ] = $scenario_post_data[ 'scenario_actions' ][ 'conditional_components' ][ 'hidden_components' ];
		}

		return $scenario_meta;
	}

	/**
	 * Front-end script.
	 *
	 * @param array $dependencies
	 */
	public static function add_to_cart_script( $dependencies ) {

		$suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';

		wp_register_script( 'wccp-conditional-components', self::plugin_url() . '/assets/js/wccp-conditional-components-add-to-cart' . $suffix . '.js', array(), self::$version );

		$dependencies[] = 'wccp-conditional-components';

		return $dependencies;
	}

	/**
	 * Admin writepanel scripts.
	 *
	 * @return void
	 */
	public static function admin_scripts() {

		$suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';

		$writepanel_dependency = 'wc_composite_writepanel';

		wp_register_script( 'wccp-conditional-components-writepanel', self::plugin_url() . '/assets/js/wccp-conditional-components-write-panels' . $suffix . '.js', array( 'jquery', 'jquery-ui-datepicker', $writepanel_dependency ), self::$version );
		wp_register_style( 'wccp-conditional-components-writepanel-css', self::plugin_url() . '/assets/css/wccp-conditional-components-write-panels.css', array( 'woocommerce_admin_styles' ), self::$version );

		// Get admin screen id
		$screen = get_current_screen();

		// WooCommerce admin pages
		if ( in_array( $screen->id, array( 'product' ) ) ) {
			wp_enqueue_script( 'wccp-conditional-components-writepanel' );
		}

		if ( in_array( $screen->id, array( 'edit-product', 'product' ) ) ) {
			wp_enqueue_style( 'wccp-conditional-components-writepanel-css' );
		}

	}

	/**
	 * Add 'Hide Components' action in Scenarios.
	 *
	 * @return void
	 */
	public static function actions_config( $id, $scenario_data, $composite_data, $product_id ) {

		$hide_components   = isset( $scenario_data[ 'scenario_actions' ][ 'conditional_components' ][ 'is_active' ] ) ? $scenario_data[ 'scenario_actions' ][ 'conditional_components' ][ 'is_active' ] : 'no';
		$hidden_components = ! empty( $scenario_data[ 'scenario_actions' ][ 'conditional_components' ][ 'hidden_components' ] ) ? $scenario_data[ 'scenario_actions' ][ 'conditional_components' ][ 'hidden_components' ] : array();

		?>
		<div class="scenario_action_conditional_components_group" >
			<div class="form-field toggle_conditional_components">
				<label for="scenario_action_conditional_components_<?php echo $id; ?>">
					<?php echo __( 'Hide components', 'woocommerce-composite-products-conditional-components' ); ?>
				</label>
				<input type="checkbox" class="checkbox" <?php echo ( $hide_components === 'yes' ? ' checked="checked"' : '' ); ?> name="bto_scenario_data[<?php echo $id; ?>][scenario_actions][conditional_components][is_active]" <?php echo ( $hide_components === 'yes' ? ' value="1"' : '' ); ?> />
				<?php echo WC_CP_Core_Compatibility::wc_help_tip( __( 'Hide one or more Components when this Scenario is active.', 'woocommerce-composite-products-conditional-components' ) ); ?>
			</div>
			<div class="action_components" <?php echo ( $hide_components === 'no' ? ' style="display:none;"' : '' ); ?> >
				<select id="bto_conditional_components_ids_<?php echo $id; ?>" name="bto_scenario_data[<?php echo $id; ?>][scenario_actions][conditional_components][hidden_components][]" style="width: 75%;" class="<?php echo WC_CP_Core_Compatibility::is_wc_version_gte_2_3() ? 'wc-enhanced-select' : 'chosen_select'; ?> conditional_components_ids" multiple="multiple" data-placeholder="<?php echo __( 'Select components&hellip;', 'woocommerce-composite-products-conditional-components' ); ?>"><?php

					foreach ( $composite_data as $component_id => $component_data ) {

						$component_title = apply_filters( 'woocommerce_composite_component_title', $component_data[ 'title' ], $component_id, $product_id );

						$option_selected = in_array( $component_id, $hidden_components ) ? 'selected="selected"' : '';
						echo '<option ' . $option_selected . 'value="' . $component_id . '">' . $component_title . '</option>';
					}

				?></select>
			</div>
		</div>
		<?php
	}

}

WC_CP_Scenario_Action_Conditional_Components::init();
